### PR TITLE
Implements one-tap links

### PR DESCRIPTION
**Pull request by wassgha**
Partial for #253 

ezgif com-video-to-gif (22)

### Changes
- Adds ability to heuristically infer link type (two-tap/one-tap) based on the page and placement of the element
- Implements rendering of CTA links in an `<amp-story-cta-layer>` of the output result
- Adds visual indicator for the one-tap vs two-tap links areas

### CODE REVIEW

File: assets/src/edit-story/components/canvas/singleSelectionMovable.js [hunk 1]
```diff
@@ -122,6 +122,8 @@ function SingleSelectionMovable({ selectedElement, targetEl, pushEvent }) {
     }
     pushTransform(selectedElement.id, frame);
   };
+  // TODO(@wassgha): Figure out why this is needed? Should be already bound
```
Comment (dvoytenko) [inline assets/src/edit-story/components/canvas/singleSelectionMovable.js]: We need to figure this out. There's no `this` here technically. So it's rather surprising.

File: assets/src/edit-story/components/link/guidelines.js [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+import { LinkType, inferLinkType, getLinkFromElement } from '.';
+
+const Separator = styled.div`
+  position: absolute;
+  bottom: ${CTA_ZONE_PERCENT * 100}%;
+  left: 0px;
+  width: 100%;
+  border-bottom: 3px dashed ${({ theme }) => rgba(theme.colors.fg.v1, 0.4)};
+  mix-blend-mode: difference;
+  z-index: 1;
```
Comment (dvoytenko) [inline assets/src/edit-story/components/link/guidelines.js:45]: `pointer-events: none`

File: assets/src/edit-story/components/link/guidelines.js [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    if (hasOneTapLinks) {
+      setLinkType(inferLinkType(selectedElement?.y));
+    }
+  }, [hasOneTapLinks, selectedElement, setLinkType]);
+
+  useTransformHandler(selectedElement?.id, (transform) => {
+    if (!hasOneTapLinks) {
+      return;
+    }
+    const translateY = transform?.translate?.[1];
```
Comment (dvoytenko) [inline assets/src/edit-story/components/link/guidelines.js:95]: `x` matters too, right? e.g. there's a forbidden zone for links?

File: assets/src/edit-story/components/link/guidelines.js [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    actions: { editorToDataY },
+  } = useUnits();
+
+  const selectedElement = selectedElements.length === 1 && selectedElements[0];
+  const link = getLinkFromElement(selectedElement);
+  const hasOneTapLinks = selectedElement && link && currentPageNumber !== 1;
+
+  useEffect(() => {
+    if (hasOneTapLinks) {
+      setLinkType(inferLinkType(selectedElement?.y));
```
Comment (dvoytenko) [inline assets/src/edit-story/components/link/guidelines.js]: Hmm. I'm afraid it's not so simple. But I'd be ok if this will go into a separate pull request with a filed issue. But I believe, we need at minimum `y`, `height` and `rotationAngle` to calculate this value. Basically, we need to do whatever Moveable does to calculate the complete outline.

File: assets/src/edit-story/components/link/guidelines.js [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    if (!hasOneTapLinks) {
+      return;
+    }
+    const translateY = transform?.translate?.[1];
+    if (translateY) {
+      setLinkType(inferLinkType(selectedElement.y + editorToDataY(translateY)));
+    }
+  });
+
+  useEffect(() => {
```
Comment (dvoytenko) [inline assets/src/edit-story/components/link/guidelines.js:108]: Are the any risks to do this immediately, instead of waiting for `dragEnd`? We've been reserving an ability to cancel the drag operation. And we do have some form of "completion" precedent that we use for resize. Likewise, I'm ok if we start here, and improve if needed. But let's check-in on this thread for cons/pros.

File: assets/src/edit-story/components/panels/link.js [hunk 6]
```diff
@@ -53,10 +53,14 @@ const Note = styled.span`
 `;
 
 function LinkPanel({ selectedElements, onSetProperties }) {
+  const y = getCommonValue(selectedElements, 'y');
```
Comment (dvoytenko) [inline assets/src/edit-story/components/panels/link.js]: Missed this before, but the `const link = getCommonValue(..., 'link')` - that would never yield "common value" since it's a freely-created object.

File: assets/src/edit-story/components/panels/link.js [hunk 7]
```diff
@@ -53,10 +53,14 @@ const Note = styled.span`
...
   const link = getCommonValue(selectedElements, 'link') || null;
   const isFill = getCommonValue(selectedElements, 'isFill');
+  const inferredLinkType = inferLinkType(y);
```
Comment (dvoytenko) [inline assets/src/edit-story/components/panels/link.js]: Just to confirm: is this solely for the "new link" case? And in either case, let's use `useMemo()` here - that math will be getting more nuanced.

File: assets/src/edit-story/components/link/index.js [hunk 8]
```diff
@@ -23,6 +30,12 @@ export function getLinkFromElement(element) {
   return element.link || null;
 }
 
+export function inferLinkType(y) {
```
Comment (dvoytenko) [inline assets/src/edit-story/components/link/index.js]: IMHO, we're better off to just always pass the whole element here as an arg or an open object - the number of args is likely to grow. Also, pls add jsdoc.

File: assets/src/edit-story/constants.js [hunk 9]
```diff
@@ -40,6 +40,11 @@ export const DEFAULT_EM = PAGE_HEIGHT * 0.02186;
 export const MIN_FONT_SIZE = 30;
 export const MAX_FONT_SIZE = 200;
 
+export const CTA_ZONE_PERCENT = 0.2; /** Bottom 20% */
```
Comment (dvoytenko) [inline assets/src/edit-story/constants.js:45]: Do we have any reference to all four of these values somewhere in the Stories docs?

File: assets/src/edit-story/output/element.js [hunk 10]
```diff
@@ -22,8 +27,9 @@ import WithMask from '../masks/output';
 import { getDefinitionForType } from '../elements';
 import { getBox } from '../units/dimensions';
 import WithLink from '../components/link/output';
+import { CTA_ZONE_PERCENT } from '../constants';
 
-function OutputElement({ element }) {
+function OutputElement({ element, cta }) {
```
Comment (dvoytenko) [inline assets/src/edit-story/output/element.js]: Do we really need to pass `cta` here? The `element.link.type` already has that info, right? Though if you feel strongly about this - that's ok.

File: assets/src/edit-story/output/element.js [hunk 11]
```diff
@@ -41,9 +47,9 @@ function OutputElement({ element }) {
       className="wrapper"
       style={{
         left: `${x}%`,
-        top: `${y}%`,
+        top: `${cta ? y - 100 * (1 - CTA_ZONE_PERCENT) : y}%`,
```
Comment (dvoytenko) [inline assets/src/edit-story/output/element.js]: Would it be better if we were to modify `box` calculation above? The box is passed around and I'm wondering if that could be an issue with it having the incorrect `height` in the sub-components. Also, the opposite could be true. So, please check what's better.

File: assets/src/edit-story/output/page.js [hunk 12]
```diff
@@ -67,12 +80,20 @@ function OutputPage({ page }) {
           ))}
         </div>
       </amp-story-grid-layer>
+      {ctaElements.length && index >= 1 && (
```
Comment (dvoytenko) [inline assets/src/edit-story/output/page.js]: I think the `index >= 1` rule should be enforced on the data model level, not here.

**Review by dvoytenko — CHANGES_REQUESTED** [review]

File: assets/src/edit-story/components/link/guidelines.js (hunk 3 above)
Comment (wassgha) [reply assets/src/edit-story/components/link/guidelines.js:95]: leaving that to the edge cases PR after talking with Sam

File: assets/src/edit-story/components/link/guidelines.js (hunk 4 above)
Comment (wassgha) [reply assets/src/edit-story/components/link/guidelines.js]: Yup, will add on the edge cases PR

File: assets/src/edit-story/components/link/guidelines.js (hunk 3 above)
Comment (dvoytenko) [reply assets/src/edit-story/components/link/guidelines.js:95]: Sure! but pls file an issue.

File: assets/src/edit-story/components/link/guidelines.js (hunk 5 above)
Comment (wassgha) [reply assets/src/edit-story/components/link/guidelines.js:108]: I think this actually gives the user good instant feedback, if our effect/callback dependency arrays are set up correctly this shouldn't have any bad consequences and would actually be a good test for how well scoped those dependency arrays are, see below
ezgif com-video-to-gif (27)

Comment (dvoytenko) [reply assets/src/edit-story/components/link/guidelines.js:108]: Ok. Let's continue down this path then.

File: package.json [hunk 13]
```diff
@@ -157,6 +157,9 @@
       "pre-commit": "lint-staged"
     }
   },
+  "resolutions": {
```
Comment (spacedmonkey) [inline package.json]: What does this do?

Comment (swissspidy) [reply package.json]: That seems to be for Yarn's [selective dependency resolutions](https://classic.yarnpkg.com/en/docs/selective-version-resolutions/) feature. But we use npm, which does not support that. Unless we'd use a package like https://www.npmjs.com/package/npm-force-resolutions.

The question is: what kind of issues do we have with the regular inferred version of the scheduler package?

File: package.json [hunk 14]
```diff
@@ -115,6 +115,7 @@
     "prebuild:js": "rm -rf assets/css/* assets/js/*",
     "build:js": "wp-scripts build",
     "build:plugin": "wp dist-archive . $(pwd)/build/web-stories --create-target-dir",
+    "build:storybook": "build-storybook -c .storybook -o build/storybook --quiet",
```
Comment (swissspidy) [inline package.json:122]: That means `npm run build` will also build the storybook which might be unexpected. Otherwise this works for me.

File: assets/src/edit-story/components/link/guidelines.js (hunk 3 above)
Comment (wassgha) [reply assets/src/edit-story/components/link/guidelines.js:95]: Filed https://github.com/google/web-stories-wp/issues/646

File: assets/src/edit-story/components/link/guidelines.js (hunk 4 above)
Comment (wassgha) [reply assets/src/edit-story/components/link/guidelines.js]: https://github.com/google/web-stories-wp/issues/646

File: assets/src/edit-story/output/element.js (hunk 11 above)
Comment (wassgha) [reply assets/src/edit-story/output/element.js]: I don't think so, the box calculation doesn't need to be skewed the wrapper is really an edge case here

File: package.json (hunk 13 above)
Comment (wassgha) [reply package.json]: I was just debugging an issue and it seemed like resolving a conflict between 0.18 and 0.19 coming from two different packages was a potential solution but it turned out to be a different problem. I removed this for now but let's keep an eye on it.

File: package.json (hunk 14 above)
Comment (wassgha) [reply package.json:122]: I think that's alright

File: assets/src/edit-story/output/element.js (hunk 11 above)
Comment (dvoytenko) [reply assets/src/edit-story/output/element.js]: Sounds good.

File: assets/src/edit-story/components/canvas/singleSelectionMovable.js (hunk 1 above)
Comment (dvoytenko) [reply assets/src/edit-story/components/canvas/singleSelectionMovable.js]: ping on this

Comment (wassgha) [reply assets/src/edit-story/components/canvas/singleSelectionMovable.js]: Fixed

**Comment by wassgha** [comment]
@dvoytenko let's merge this?

File: assets/src/edit-story/components/link/guidelines.js [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+
+const Separator = styled.div`
+  position: absolute;
+  bottom: ${CTA_ZONE_PERCENT * 100}%;
+  left: 0px;
+  width: 100%;
+  mix-blend-mode: difference;
+  pointer-events: none;
+  z-index: 1;
+  background-image: linear-gradient(
```
Comment (swissspidy) [inline assets/src/edit-story/components/link/guidelines.js:46]: Why isn't this just a dashed border like on Figma? Seems overly complicated.

File: assets/src/edit-story/components/link/guidelines.js [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+const Tip = styled.span`
+  position: absolute;
+  left: 100%;
+  white-space: nowrap;
+  margin-left: 12px;
+  font-size: 12px;
+  filter: none;
+  color: ${({ theme, active }) =>
+    active ? theme.colors.success.v0 : rgba(theme.colors.fg.v1, 0.84)};
+  ${({ pos }) => (pos === 'top' ? `bottom: 0px;` : `top: 0px;`)}
```
Comment (swissspidy) [inline assets/src/edit-story/components/link/guidelines.js]: The distance between the tips seems to be a bit bigger on Figma

File: assets/src/edit-story/components/link/guidelines.js (hunk 15 above)
Comment (wassgha) [reply assets/src/edit-story/components/link/guidelines.js:46]: Sam didn't like it and I agree, the distance between the dashes on the default browser dashed border doesn't look good so I had to improvise

**Comment by swissspidy** [comment]
Hmm when dragging an element in the bottom 20% of the page there is quite a noticeable lag. Even when dragging an element on the first page, where there are no 1-tap links allowed.

Also, I cannot drag an element from the top part to the bottom 20%, it jumps right back. Same when it is already in the bottom 20%: I cannot move it upwards, it jumps right back. It works when using the arrow keys though.

File: assets/src/edit-story/components/link/guidelines.js (hunk 15 above)
Comment (swissspidy) [reply assets/src/edit-story/components/link/guidelines.js:46]: Just tested it and have to agree with it 👍

**Comment by swissspidy** [comment]
Oh, and the percentage-based font size doesn't seem work within the CTA layer. All text elements are displayed super small on the front end.

**Comment by wassgha** [comment]
> Hmm when dragging an element in the bottom 20% of the page there is quite a noticeable lag. Even when dragging an element on the first page, where there are no 1-tap links allowed.
> 
> Also, I cannot drag an element from the top part to the bottom 20%, it jumps right back. Same when it is already in the bottom 20%: I cannot move it upwards, it jumps right back. It works when using the arrow keys though.

@swissspidy both of these problems are coming from here:
https://github.com/google/web-stories-wp/blob/0bed6a74d2eac9c2735f13cd1f70e4cb77b8de36/assets/src/edit-story/components/link/guidelines.js#L82-L86
essentially, this effect only needs to be called whenever `selectedElement.y` changes (not `selectedElement` as a whole). I fixed it three times before realizing that the linter is reverting my change during the pre-submit (from `selectedElement.y` back to `selectedElement`). Any way to force it not to?

**Comment by swissspidy** [comment]
Ah, interesting!

Looks like we'll have to disable the lint rule for these specific cases. See 7d7b379bc8c0ffa35e795a8c3d282dc956fc1d24.

Also just noticed that Travis didn't fail now. Will look into that.

**Comment by swissspidy** [comment]
Works much better now with your change!

**Comment by wassgha** [comment]
Awesome thanks for that!

**Comment by swissspidy** [comment]
So the only outstanding issue now seems to be the font size one.

**Comment by wassgha** [comment]
I'm gonna try to do it in css instead of the CTA/no-cta calculations :))

**Review by dvoytenko — APPROVED** [review]
LGTM from me. But please resolve your threads with Pascal before merging.

**Comment by wassgha** [comment]
@swissspidy should be ready :)
